test/test_alloc.c

```diff
@@ -0,0 +1,570 @@
+  return 0;
+}
+
+static int test_signature_pre_hash_shake256_alloc_failure(void)
+{
+  uint8_t pk[CRYPTO_PUBLICKEYBYTES];
+  uint8_t sk[CRYPTO_SECRETKEYBYTES];
+  uint8_t sig[CRYPTO_BYTES];
+  uint8_t msg[32] = {0};
+  uint8_t rnd[32] = {0};
+  const uint8_t ctx[] = "test context";
+  size_t siglen;
+
+  reset_all();
+  if (mld_keypair(pk, sk) != 0)
+  {
+    fprintf(stderr,
+            "ERROR: mld_keypair failed in signature_pre_hash_shake256 test "
+            "setup\n");
+    return 1;
+  }
+
+  TEST_ALLOC_FAILURE(
+      "mld_signature_pre_hash_shake256",
+      mld_signature_pre_hash_shake256(sig, &siglen, msg, sizeof(msg), ctx,
+                                      sizeof(ctx) - 1, rnd, sk));
+  return 0;
+}
+
+static int test_verify_pre_hash_shake256_alloc_failure(void)
+{
+  uint8_t pk[CRYPTO_PUBLICKEYBYTES];
+  uint8_t sk[CRYPTO_SECRETKEYBYTES];
+  uint8_t sig[CRYPTO_BYTES];
+  uint8_t msg[32] = {0};
+  uint8_t rnd[32] = {0};
+  const uint8_t ctx[] = "test context";
+  size_t siglen;
+
+  reset_all();
+  if (mld_keypair(pk, sk) != 0)
+  {
+    fprintf(stderr,
+            "ERROR: mld_keypair failed in verify_pre_hash_shake256 test "
+            "setup\n");
+    return 1;
+  }
+
+  if (mld_signature_pre_hash_shake256(sig, &siglen, msg, sizeof(msg), ctx,
+                                      sizeof(ctx) - 1, rnd, sk) != 0)
+  {
+    fprintf(stderr,
+            "ERROR: mld_signature_pre_hash_shake256 failed in "
+            "verify_pre_hash_shake256 test setup\n");
+    return 1;
+  }
+
+  TEST_ALLOC_FAILURE("mld_verify_pre_hash_shake256",
+                     mld_verify_pre_hash_shake256(sig, siglen, msg, sizeof(msg),
+                                                  ctx, sizeof(ctx) - 1, pk));
+  return 0;
+}
+
+static int test_pk_from_sk_alloc_failure(void)
+{
+  uint8_t pk[CRYPTO_PUBLICKEYBYTES];
+  uint8_t sk[CRYPTO_SECRETKEYBYTES];
+
+  reset_all();
+  if (mld_keypair(pk, sk) != 0)
+  {
+    fprintf(stderr, "ERROR: mld_keypair failed in pk_from_sk test setup\n");
+    return 1;
+  }
+
+  TEST_ALLOC_FAILURE("mld_pk_from_sk", mld_pk_from_sk(pk, sk));
+  return 0;
+}
+
+int main(void)
+{
+  uint8_t bump_buffer_storage[MLD_BUMP_ALLOC_SIZE];
+  bump_buffer = bump_buffer_storage;
+
+  if (test_keygen_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_sign_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_verify_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_sign_combined_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_open_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_signature_extmu_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_verify_extmu_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_signature_pre_hash_shake256_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_verify_pre_hash_shake256_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  if (test_pk_from_sk_alloc_failure() != 0)
+  {
+    return 1;
+  }
+
+  return 0;
```

Comment:
Did you intentionally skip the following functions, or should those be added?
- crypto_sign_signature_pre_hash_internal
- crypto_sign_verify_pre_hash_internal 
- mld_prepare_domain_separation_prefix

It's probably best to add them now to not risk forgetting them in case they do add alllocations in there later.

Reply:
I left those out intentionally because they require exercising with different parameters for full coverage, and that is already achieved by the way they are called from the higher level functions?